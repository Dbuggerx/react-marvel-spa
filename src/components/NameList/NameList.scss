@import '../scss/variables';

.name-list {
  border: solid 1px map-get($colors, highlight-fg);
  margin: 5px 0;
  padding: 15px;
  width: 100%;
  height: 100%;
  box-sizing: border-box;
  background: map-get($colors, primary-bg);
  color: map-get($colors, highlight-bg);

  &__title, &__filter-input {
    margin: 5px 0;
  }

  &__title {
    color: map-get($colors, marvel-red);
    text-align: center;
  }

  &__filter-input {
    background: map-get($colors, highlight-bg);
    border: solid 1px map-get($colors, highlight-fg);
    margin-left: 5px;
    color: map-get($colors, marvel-red);
  }

  &__filter {
    display: flex;
    flex-flow: row nowrap;
    align-items: center;
    color: map-get($colors, marvel-red);
    font-size: 0.7em;
    width: 100%;
    margin-bottom: 5px;
  }

  &__item {
    margin-top: 15px;
  }
}