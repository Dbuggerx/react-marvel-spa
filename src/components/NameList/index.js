// @flow

import React, { Component } from 'react';
import './NameList.scss';

type Props = {
    names: string[],
    collectionName: string,
};

type State = {
    filterParam: string,
};

/**
 * Renders a filterable collection of strings
 */
export default class NameList extends Component<Props, State> {
    constructor(props: Props) {
        super(props);
        this.state = { filterParam: '' };
    }

    handleFilterChange = (event: SyntheticEvent<HTMLInputElement>) => {
        this.setState({
            filterParam: event.currentTarget.value,
        });
    };

    render() {
        const uniqueItems = this.props.names.reduce((acc, cur) => {
            if (acc.indexOf(cur) === -1) acc.push(cur);
            return acc;
        }, []);

        return (
            <section className="name-list">
                <h2 className="name-list__title">{this.props.collectionName}</h2>
                <aside className="name-list__filter">
                    Filter:
                    <input
                        type="text"
                        value={this.state.filterParam}
                        onChange={this.handleFilterChange}
                        className="name-list__filter-input"
                    />
                </aside>
                {(this.state.filterParam.length > 0
                    ? uniqueItems.filter(name => name.toLowerCase()
                        .indexOf(this.state.filterParam.toLowerCase()) !== -1)
                    : uniqueItems
                ).map(name => (
                    <div key={name} className="name-list__item">{name}</div>
                ))}
            </section>
        );
    }
}
