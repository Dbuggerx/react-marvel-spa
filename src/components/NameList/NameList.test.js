import React from 'react';
import { shallow } from 'enzyme';
import NameList from './index';

describe('NameList', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = shallow(<NameList names={['aaa', 'aaa', 'bbb', 'ccc']} />);
    });

    test('renders properly', () => {
        wrapper = shallow(<NameList names={['aaa', 'aaa', 'bbb', 'ccc']} />);
        expect(wrapper).toMatchSnapshot();
    });

    test('input changes state on change', () => {
        wrapper.find('input').simulate('change', {
            currentTarget: {
                value: 'test',
            },
        });

        expect(wrapper.state('filterParam')).toBe('test');
    });
});
