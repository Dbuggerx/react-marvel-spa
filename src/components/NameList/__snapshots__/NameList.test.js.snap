// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NameList renders properly 1`] = `
<section
  className="name-list"
>
  <h2
    className="name-list__title"
  />
  <aside
    className="name-list__filter"
  >
    Filter:
    <input
      className="name-list__filter-input"
      onChange={[Function]}
      type="text"
      value=""
    />
  </aside>
  <div
    className="name-list__item"
    key="aaa"
  >
    aaa
  </div>
  <div
    className="name-list__item"
    key="bbb"
  >
    bbb
  </div>
  <div
    className="name-list__item"
    key="ccc"
  >
    ccc
  </div>
</section>
`;
